src/java.base/share/classes/sun/security/provider/ML_DSA.java

```diff
@@ -0,0 +1,1427 @@
+
+        return skBytes;
+    }
+
+    public ML_DSA_PrivateKey skDecode(byte[] sk) {
+        byte[] rho = new byte[mlDsaASeedLength];
+        System.arraycopy(sk, 0, rho, 0, mlDsaASeedLength);
+
+        byte[] k = new byte[mlDsaKLength];
+        System.arraycopy(sk, mlDsaASeedLength, k, 0, mlDsaKLength);
+
+        byte[] tr = new byte[mlDsaTrLength];
+        System.arraycopy(sk, mlDsaASeedLength + mlDsaKLength, tr, 0, mlDsaTrLength);
+
+        //Parse s1
+        int start = mlDsaASeedLength + mlDsaKLength + mlDsaTrLength;
+        int end = start + (32 * mlDsa_l * s1s2CoeffSize);
+        byte[] y = Arrays.copyOfRange(sk, start, end);
+        int[][] s1 = bitUnpack(y, mlDsa_l, eta, s1s2CoeffSize);
+
+        //Parse s2
+        start = end;
+        end += 32 * s1s2CoeffSize * mlDsa_k;
+        byte[] z = Arrays.copyOfRange(sk, start, end);
+        int[][] s2 = bitUnpack(z, mlDsa_k, eta, s1s2CoeffSize);
+
+        //Parse t0
+        start = end;
+        end += 32 * mlDsa_d * mlDsa_k;
+        byte[] w = Arrays.copyOfRange(sk, start, end);
+        int[][] t0 = bitUnpack(w, mlDsa_k, 1 << 12, t0CoeffSize);
+
+        return new ML_DSA_PrivateKey(rho, k, tr, s1, s2, t0);
+    }
+
+    public byte[] sigEncode(ML_DSA_Signature sig) {
+        int cSize = lambda / 4;
+        int zSize = mlDsa_l * 32 * (1 + gamma1Bits);
+
+        byte[] zPacked = bitPack(sig.response, gamma1Bits + 1, gamma1);
+        byte[] hPacked = hintBitPack(sig.hint);
+
+        byte[] sigBytes = new byte[cSize + zSize + omega + mlDsa_k];
+        System.arraycopy(sig.commitmentHash, 0, sigBytes, 0, cSize);
+        System.arraycopy(zPacked, 0, sigBytes, cSize, zSize);
+        System.arraycopy(hPacked, 0, sigBytes, cSize + zSize, omega + mlDsa_k);
+        return sigBytes;
+    }
+
+    public ML_DSA_Signature sigDecode(byte[] sig) {
+        //Decode cTilde
+        byte[] cTilde = Arrays.copyOfRange(sig, 0, lambda/4);
+
+        //Decode z
+        int start = lambda / 4;
+        int end = start + (32 * mlDsa_l * (1 + gamma1Bits));
+        byte[] x = Arrays.copyOfRange(sig, start, end);
+        int[][] z = bitUnpack(x, mlDsa_l, gamma1, gamma1Bits + 1);
+
+        //Decode h
+        start = end;
+        end += omega + mlDsa_k;
+        byte[] y = Arrays.copyOfRange(sig, start, end);
+        boolean[][] h = hintBitUnpack(y);
+        return new ML_DSA_Signature(cTilde, z, h);
+    }
+
+    /*
+    Auxiliary functions defined in Section 7.3 of specification
+     */
+
+    private class Shake256Slicer {
+        SHAKE256 xof;
+        byte[] block;
+        int byteOffset;
+        int current;
+        int bitsInCurrent;
+        int bitsPerCall;
+        int bitMask;
+
+        Shake256Slicer(SHAKE256 xof, int bitsPerCall) {
+            this.xof = xof;
+            this.bitsPerCall = bitsPerCall;
+            bitMask = (1 << bitsPerCall) - 1;
```

Comment:
Add some comments about the limit of `bitsPerCall`, looks like cannot exceed 31.

Reply:
BitsPerCall can only be 4 (when called from sampleS1S2()) or 8 (when called from sampleInBall()), so this is not a problem.